### ISSUE TITLE
Too slow

### ISSUE THREAD

**Issue by cmp0st** [mchav/dataframe#17]
This thing runs in 5 minutes. Can we make it 2 minutes plz. Thank you :pray:

**Comment by mchav** [comment]
Right now DataFrames are defined as:

```
data DataFrame { columns :: Map Text Column, columnNames :: [Text] }
```

The main data structure is a map that stores the column names and the heterogeneous columns as values. 

Maps are implemented as tree data structures so insertion of many columns (for wide dataframes) means a lot of tree rebalancing to guarantee log n lookups. I'm not sure what sort of copying happens during the rebalance operations but it seems like it would have significant overhead.

I think moving to a vector of columns implementation might make reading from CSV faster. It may also make some operations like grouping by columns faster since vectors have better cache locality.

So we'd have to define a dataframe as:

```
data DataFrame = DataFrame {
  columns :: Vector (Maybe Column),
  columnNames :: Map Text Int,
  freeIndices :: [Int],
  size :: Int
}
```

With this implementation we can still get constant time row access but we get the data/cache locality of vectors.

The free indices and size variables will handle dynamic resizing sort of like an arraylist in Java.

I'll setup a couple of benchmarks to work against and then document the current approach's performance, Python Panda's performance and then the "improved" approach's performance.

Also worth reading and understanding [this](https://chrispenner.ca/posts/wc) to improve the string splitting logic.

**Comment by cmp0st** [comment]
Ah yes I was curious about the string splitting as it looked like a sizable contribution to read csv perf:

![here](https://private-user-images.githubusercontent.com/189583441/395764430-25e45e73-c823-4642-83c8-2a88ee170033.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyMzk4NTIsIm5iZiI6MTczNDIzOTU1MiwicGF0aCI6Ii8xODk1ODM0NDEvMzk1NzY0NDMwLTI1ZTQ1ZTczLWM4MjMtNDY0Mi04M2M4LTJhODhlZTE3MDAzMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNVQwNTEyMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZTFmNDM4ZDEyMGEzNjM4YmNmNjA1YjExZTA4MTZmZjJmZTBmZTU2ZDlhZTM0ODcyZTI1NzNjYjZhNjk5MGI4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2CYjhAIN-mex85XS8meQblmy4vfiy7VqmXPzH3GZ_Jc)

**Comment by mchav** [comment]
I have a separate branch where I try and read columns into the target type instead of reading text and parsing later.

The work for this is in the [better reads branch](https://github.com/mchav/dataframe/tree/better_reads). The rough approach was:

* Guess the type from the first element
* Initialize vectors of `Maybe a` where `a` is each inferred type.
* Parse the data on read  then write it into a `MutableColumn` which is a wrapper around a mutable vector.
* Unwrap the `Maybe a` into an `a` then unbox the vector if necessary.

This turned out to speed up the execution but only very slightly. I used the credit card data from Kaggle:

```
$ cp ../datasets/creditcard.csv ../datasets/augmented.csv
$ sed -i '1d' ../datasets/augmented.csv 
$ less ../datasets/augmented.csv 
$ cp ../datasets/creditcard.csv ../datasets/augmented_final.csv
$ cat ../datasets/augmented.csv >>^C
$ cat ../datasets/augmented.csv >> ../datasets/augmented_final.csv
```

I ran both versions on the initial credit card data file then on the augmented version that's double the size (`augmented_final.csv`).

The actual code looks like:

```haskell
df <- D.readCsv "<filename>"
print $ D.take 10
```

Current implementation on the singular 140MB file:
![current_implementation_single](https://github.com/user-attachments/assets/f80c61f9-ec18-465b-9870-e2a49f9cd70d)

New implementation on the singular 140MB file:
![new_implementation_single](https://github.com/user-attachments/assets/50d0818b-35f2-4d7b-8ce1-edd505af1302)

Current implementation on doubled 280MB file:
![current_implementation_double](https://github.com/user-attachments/assets/533618f1-6b8c-4b8e-8b4b-a590829fee3c)


New implementation on doubled 280MB file:
![new_implementation_double](https://github.com/user-attachments/assets/202ea184-ac6a-4760-a6be-905b059455f4)


So we get a 15% speed decrease but the memory footprint of the code doubles. I think it's probably better to have a smaller memory footprint and increase speed through parallelization since Haskell is a memory hog. 

Looking further into the memory profile it seems the main problem comes from parsing eagerly. The new implementation has a very heavy `writeColumn`. The current implementation amortizes by being lazy about it. So rows aren't really parsed until they are needed thus spreading the memory usage out.

Current implementation 280MB file memory profile:
![current_double_memory](https://github.com/user-attachments/assets/68afc64f-7a2e-44a4-a185-753d417ede3f)

New implementation 280MB file memory profile:
![new_double_memory](https://github.com/user-attachments/assets/8a327d9c-208e-4521-b7e4-c7eaa68acdd3)

In short the speedup from eager parsing is ~15% and the memory footprint increase is ~150%. That doesn't seem like a good deal.

But this profiling taught me two things:
* Our optimizations should depend on expected usage patterns - I think absolute benchmark optimizations might involve tradeoffs that I can reason about but may be contrary to users' needs. A good place to start looking into that direction is modin's [opportunistic evaluation](https://arxiv.org/pdf/2103.02145) model.
* There is room for a parallelism in the library.

I'll keep documenting these experiments under this issue.

**Comment by mchav** [comment]
Initially experiments on parallelism weren't too promising.

```haskell
-- | Reads rows from the handle and stores values in mutable vectors.
fillColumns :: Int -> Char -> MVar (VM.IOVector (VM.IOVector T.Text)) -> Handle -> IO ()
fillColumns i c mutableCols handle = do
    isEOF <- hIsEOF handle
    unless isEOF $ do
        row <- split c <$> TIO.hGetLine handle
        mapConcurrently_(uncurry (writeValue mutableCols i)) (zip [0..] row)
        fillColumns (i + 1) c mutableCols handle
```

and

```haskell
cols <- mapConcurrently (freezeColumn mColVar opts) (V.generate numColumns id)
```

Sync
![sync](https://github.com/user-attachments/assets/24fcc4cc-75f2-4422-b8b6-a3f1c7b68932)

Async
![async](https://github.com/user-attachments/assets/aa782293-809f-4862-984c-65d70d25c69a)

**Comment by mchav** [comment]
Removing some extra copies when we are parsing (i.e the conversion copies) helped take our running time down by a lot. It seems we can get a lot of mileage from addressing unnecessary traversals.

Before removing parsing traversal:
![parse_copies](https://github.com/user-attachments/assets/51c49545-d987-41e0-92c8-0b9a4c01eb28)

After:
![reduced_copies](https://github.com/user-attachments/assets/a4717ab4-f07d-4850-b77f-0bc2acbdf39f)

**Comment by mchav** [comment]
> Initially experiments on parallelism weren't too promising.
> 
> ```haskell
> -- | Reads rows from the handle and stores values in mutable vectors.
> fillColumns :: Int -> Char -> MVar (VM.IOVector (VM.IOVector T.Text)) -> Handle -> IO ()
> fillColumns i c mutableCols handle = do
>     isEOF <- hIsEOF handle
>     unless isEOF $ do
>         row <- split c <$> TIO.hGetLine handle
>         mapConcurrently_(uncurry (writeValue mutableCols i)) (zip [0..] row)
>         fillColumns (i + 1) c mutableCols handle
> ```
> 
> and
> 
> ```haskell
> cols <- mapConcurrently (freezeColumn mColVar opts) (V.generate numColumns id)
> ```
> 
> Sync ![sync](https://private-user-images.githubusercontent.com/10537221/399185121-24fcc4cc-75f2-4422-b8b6-a3f1c7b68932.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1MzYzOTAsIm5iZiI6MTczNTUzNjA5MCwicGF0aCI6Ii8xMDUzNzIyMS8zOTkxODUxMjEtMjRmY2M0Y2MtNzVmMi00NDIyLWI4YjYtYTNmMWM3YjY4OTMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDA1MjEzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwNWI0MTU5OWNhYzQ4NDZmMjIyZDQ3OTdmMDE1NGYzZjZmZDE2Yjg5NDU1Y2ZmOWIxNzVkMzY4MGZmMTcxZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1Yora9HH5abyWn62Kto--9jb-xqjTCPpC7d3Qfh7PZA)
> 
> Async ![async](https://private-user-images.githubusercontent.com/10537221/399185126-aa782293-809f-4862-984c-65d70d25c69a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1MzYzOTAsIm5iZiI6MTczNTUzNjA5MCwicGF0aCI6Ii8xMDUzNzIyMS8zOTkxODUxMjYtYWE3ODIyOTMtODA5Zi00ODYyLTk4NGMtNjVkNzBkMjVjNjlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDA1MjEzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhYTYxMTkwZWU3MjUwNWUwOTdlZDhkYTNmOGU0Nzk5ZmVjZDk2MjZkYjg5ZjAwNjFiMmY3OTA5NmJmYmEyNWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.H0riXqxJkQhTKaj3czrHisjqroAuiDhEgzn6F_9cZKY)

Actually this was a bad experiment since I didn't compile with `-threaded`. When compiling with `-threaded` the memory profile doubles. The time in the profile doubles as well but the time with profiling disabled remains the same.

**Comment by mchav** [comment]
I revisited the better reads approach. The initial implementation's bottleneck was reading everything into Vectors of Text. We only parsed after all the values had been accumulated. The new approach does the following:

* Read the first row and "guess" the types of each column.
* Assign a mutable boxed or unboxed column based on the inferred type.
* Create a `Vector (Set Int)` for each column which tracks if we encountered a null.
* Read the file, split the row and put all the rows into the relevant type. If we encounter a nullish value then store 0/empty string and then update the nulls vector set with the integer of the null position.
* After freezing we check if the column we just froze had nulls. If it does replace the nullish indices with Nothing and the non nulls with `Just value`

This SIGNIFICANTLY reduces our memory footprint:

For the same benchmark in the previous comments:

![final_profile](https://github.com/user-attachments/assets/5aefe825-4913-4fa6-ab81-8557f488754b)

Running time wise this reduced our initial file loading time from ~5 minutes to ~20 seconds. Still 6 times slower than Python's read_csv (runs in 3s on my machine for the same dataset) but much better than our original time.

**Comment by mchav** [comment]
It might be worth fundamentally rethinking how to parse CSV. I'll spend some time looking at the python CSV module.

**Comment by mchav** [comment]
When compiled and installed the library's performance comes close to the Polars' (6 seconds vs 3 seconds).

**Comment by mchav** [comment]
This is probably as fast as CSV reading will get. With the recent change to read multi-line CSV (where a row might span over multiple lines) it seems performance degrades to ~9 seconds. So still much slower than Polars but outside of rolling up a custom text implementation I'm not sure what else to do. I'll open more targeted issues later as performance concerns come up.
